File: BraintreeModels/src/main/java/com/braintreepayments/api/DeviceInspector.java
package com.braintreepayments.api;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.os.Build;

import androidx.annotation.VisibleForTesting;

import java.io.BufferedReader;
import java.io.File;
import java.io.InputStreamReader;

public class DeviceInspector {

    private static final String PAYPAL_APP_PACKAGE = "com.paypal.android.p2pmobile";
    private static final String VENMO_APP_PACKAGE = "com.venmo";

    private static final String VENMO_APP_SWITCH_ACTIVITY = "controller.SetupMerchantActivity";
    private static final String VENMO_CERTIFICATE_SUBJECT = "CN=Andrew Kortina,OU=Engineering,O=Venmo,L=Philadelphia,ST=PA,C=US";
    private static final String VENMO_CERTIFICATE_ISSUER = "CN=Andrew Kortina,OU=Engineering,O=Venmo,L=Philadelphia,ST=PA,C=US";

    private static final int VENMO_PUBLIC_KEY_HASH_CODE = -129711843;

    private final AppHelper appHelper;
    private final ManifestValidator manifestValidator;

    public DeviceInspector() {
        this(new AppHelper(), new ManifestValidator());
    }

    @VisibleForTesting
    DeviceInspector(AppHelper appHelper, ManifestValidator manifestValidator) {
        this.appHelper = appHelper;
        this.manifestValidator = manifestValidator;
    }

    public boolean isPayPalInstalled(Context context) {
        return appHelper.isAppInstalled(context, PAYPAL_APP_PACKAGE);
    }

    public boolean isVenmoInstalled(Context context) {
        return appHelper.isAppInstalled(context, VENMO_APP_PACKAGE);
    }

    /**
     * @param context A context to access the installed packages.
     * @return boolean depending on if the Venmo app is installed, and has a valid signature.
     */
    public boolean isVenmoAppSwitchAvailable(Context context) {
        return appHelper.isIntentAvailable(context, getVenmoIntent()) &&
                SignatureVerification.isSignatureValid(context, VENMO_APP_PACKAGE, VENMO_CERTIFICATE_SUBJECT, VENMO_CERTIFICATE_ISSUER,
                        VENMO_PUBLIC_KEY_HASH_CODE);
    }

    private static Intent getVenmoIntent() {
        return new Intent().setComponent(new ComponentName(VENMO_APP_PACKAGE, VENMO_APP_PACKAGE + "." + VENMO_APP_SWITCH_ACTIVITY));
    }

    public boolean isDeviceEmulator() {
        return isDeviceEmulator(Build.PRODUCT, Build.MANUFACTURER, Build.FINGERPRINT);
    }

    @VisibleForTesting
    boolean isDeviceEmulator(String buildProduct, String buildManufacturer, String buildFingerprint) {
        if ("google_sdk".equalsIgnoreCase(buildProduct) ||
                "sdk".equalsIgnoreCase(buildProduct) ||
                "Genymotion".equalsIgnoreCase(buildManufacturer) ||
                buildFingerprint.contains("generic")) {
            return true;
        } else {
            return false;
        }
    }

    public String getAppName(Context context) {
        ApplicationInfo applicationInfo;
        String packageName = context.getPackageName();
        PackageManager packageManager = context.getPackageManager();
        try {
            applicationInfo = packageManager.getApplicationInfo(packageName, 0);
        } catch (PackageManager.NameNotFoundException e) {
            applicationInfo = null;
        }

        String appName = null;
        if (applicationInfo != null) {
            appName = (String) packageManager.getApplicationLabel(applicationInfo);
        }

        if (appName == null) {
            return "ApplicationNameUnknown";
        }
        return appName;
    }

    public boolean isDeviceRooted() {
        return isDeviceRooted(
                android.os.Build.TAGS, new File("/system/app/Superuser.apk"), Runtime.getRuntime());
    }

    @VisibleForTesting
    boolean isDeviceRooted(String buildTags, File superUserApkFile, Runtime runtime) {
        boolean check1 = buildTags != null && buildTags.contains("test-keys");

        boolean check2;
        try {
            check2 = superUserApkFile.exists();
        } catch (Exception e) {
            check2 = false;
        }

        boolean check3;
        try {
            Process process = runtime.exec(new String[]{"/system/xbin/which", "su"});
            BufferedReader in = new BufferedReader(new InputStreamReader(process.getInputStream()));
            check3 = in.readLine() != null;
        } catch (Exception e) {
            check3 = false;
        }

        return (check1 || check2 || check3);
    }
}

File: BraintreeCore/src/main/java/com/braintreepayments/api/core/DeviceInspector.kt
package com.braintreepayments.api.core

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import androidx.annotation.RestrictTo
import com.braintreepayments.api.sharedutils.AppHelper
import com.braintreepayments.api.sharedutils.SignatureVerifier

/**
 * @suppress
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
class DeviceInspector(
    private val appHelper: AppHelper = AppHelper(),
    private val signatureVerifier: SignatureVerifier = SignatureVerifier(),
) {

    internal fun getDeviceMetadata(
        context: Context?,
        configuration: Configuration?,
        sessionId: String?,
        integration: IntegrationType?
    ): DeviceMetadata {
        return DeviceMetadata(
            appId = context?.packageName,
            appName = getAppName(context),
            clientSDKVersion = BuildConfig.VERSION_NAME,
            clientOs = getAPIVersion(),
            component = "braintreeclientsdk",
            deviceManufacturer = Build.MANUFACTURER,
            deviceModel = Build.MODEL,
            dropInSDKVersion = dropInVersion,
            environment = configuration?.environment,
            eventSource = "mobile-native",
            integrationType = integration,
            isSimulator = isDeviceEmulator,
            merchantAppVersion = getAppVersion(context),
            merchantId = configuration?.merchantId,
            platform = "Android",
            sessionId = sessionId
        )
    }

    // Analytics payload no longer sends appInstalled info.
    // Leaving logic for upcoming PaymentReady API implementation.
    /**
     * @param context A context to access the installed packages.
     * @return boolean depending on if the Venmo app is installed, and has a valid signature.
     */
    fun isVenmoAppSwitchAvailable(context: Context?): Boolean {
        val isVenmoIntentAvailable = appHelper.isIntentAvailable(context, venmoIntent)
        val isVenmoSignatureValid = signatureVerifier.isSignatureValid(
            context, VENMO_APP_PACKAGE, VENMO_BASE_64_ENCODED_SIGNATURE
        )
        return isVenmoIntentAvailable && isVenmoSignatureValid
    }

    fun isPayPalInstalled(context: Context?): Boolean {
        return appHelper.isAppInstalled(context, PAYPAL_APP_PACKAGE)
    }

    fun isPayPalBetaInstalled(context: Context?): Boolean {
        return appHelper.isAppInstalled(context, PAYPAL_APP_PACKAGE_BETA)
    }

    fun isVenmoInstalled(context: Context?): Boolean {
        return appHelper.isAppInstalled(context, VENMO_APP_PACKAGE)
    }

    private val isDeviceEmulator: Boolean
        get() = Build.BRAND.startsWith("generic") &&
            Build.DEVICE.startsWith("generic") ||
            Build.FINGERPRINT.startsWith("generic") ||
            Build.FINGERPRINT.startsWith("unknown") ||
            Build.HARDWARE.contains("goldfish") ||
            Build.HARDWARE.contains("ranchu") ||
            Build.MODEL.contains("google_sdk") ||
            Build.MODEL.contains("Emulator") ||
            Build.MODEL.contains("Android SDK built for x86") ||
            Build.MANUFACTURER.contains("Genymotion") ||
            Build.PRODUCT.contains("sdk_google") ||
            Build.PRODUCT.contains("google_sdk") ||
            Build.PRODUCT.contains("sdk") ||
            Build.PRODUCT.contains("sdk_x86") ||
            Build.PRODUCT.contains("vbox86p") ||
            Build.PRODUCT.contains("emulator") ||
            Build.PRODUCT.contains("simulator")

    private fun getAppName(context: Context?): String =
        getApplicationInfo(context)?.let { appInfo ->
                context?.packageManager?.getApplicationLabel(appInfo).toString()
        } ?: "ApplicationNameUnknown"

    @Suppress("SwallowedException")
    private fun getApplicationInfo(context: Context?) =
        try {
            context?.packageManager?.getApplicationInfo(context.packageName, 0)
        } catch (e: PackageManager.NameNotFoundException) {
            null
        }

    private fun getAppVersion(context: Context?): String = getPackageInfo(context) ?: "VersionUnknown"

    private fun getPackageInfo(context: Context?) =
        context?.let {
            try {
                val packageInfo = it.packageManager.getPackageInfo(it.packageName, 0)
                packageInfo?.versionName
            } catch (ignored: PackageManager.NameNotFoundException) { null }
        }

    private fun getAPIVersion(): String {
        val sdkInt = Build.VERSION.SDK_INT
        return "Android API $sdkInt"
    }

    /**
     * Gets the current Drop-in version or null.
     *
     * @return string representation of the current Drop-in version, or null if
     * Drop-in is unavailable
     */
    private val dropInVersion
        get() = getDropInVersion()

    companion object {
        private const val PAYPAL_APP_PACKAGE = "com.paypal.android.p2pmobile"
        private const val PAYPAL_APP_PACKAGE_BETA = "com.paypal.android.p2pmobile.pp_beta"
        private const val VENMO_APP_PACKAGE = "com.venmo"
        private const val VENMO_APP_SWITCH_ACTIVITY = "controller.SetupMerchantActivity"

        const val VENMO_BASE_64_ENCODED_SIGNATURE = "x34mMawEUcCG8l95riWCOK+kAJYejVmdt44l6tzcyUc=\n"
        private val venmoIntent: Intent
            get() = Intent().setComponent(
                ComponentName(
                    VENMO_APP_PACKAGE,
                    "$VENMO_APP_PACKAGE.$VENMO_APP_SWITCH_ACTIVITY"
                )
            )
        internal fun getDropInVersion(): String? {
            try {
                val dropInBuildConfigClass = Class.forName("com.braintreepayments.api.dropin.BuildConfig")
                val versionNameField = dropInBuildConfigClass.getField("VERSION_NAME")
                versionNameField.isAccessible = true
                return versionNameField[String::class] as String?
            } catch (ignored: ClassNotFoundException) {
            } catch (ignored: NoSuchFieldException) {
            } catch (ignored: IllegalAccessException) {
            }
            return null
        }
    }
}
